modules/vslam/slam.cpp

```diff
@@ -0,0 +1,506 @@
+		else 
+			RTFromFundamental(tform, points1, points2, rotations, translations);
+
+		std::vector<Point3f> points_3f;
+		int solution = motionChooseSolution(points1, points2, rotations, translations, points_3f);
+
+		R.push_back(rotations[solution]);
+		t.push_back(translations[solution]);
+		visualizeStructureAndMotion(points_3f, R, t);
+		point_3f_frame.push_back(points_3f);
+		frames_number += 2;
+		is_map_init = true;
+		return true;
+	}
+	else if (is_map_init)
+	{
+		std::cout << "Map is already init" << std::endl;
+		return false;
+	}
+	else
+	{
+		std::cout << "Need two images to start initialization" << std::endl;
+		return false;
+	}
+}
+
+bool Vslam::tracking(const Mat& img1, const Mat& img2)
+{
+	if ((is_map_init) && (!img2.empty()))
+	{
+		std::vector<KeyPoint> prev_features_tmp = curr_features;
+		prev_descriptors = curr_descriptors.clone();
+
+		std::vector<KeyPoint> curr_features_tmp;
+		Mat curr_descriptors_tmp;
+
+		detectFeatures(img2, curr_features_tmp, curr_descriptors_tmp);
+	
+		std::vector<DMatch> matches;
+		std::vector<int> matched_ind1, matched_ind2, unmatched_ind1, unmatched_ind2;
+
+		if (!matchFeatures(prev_descriptors, curr_descriptors_tmp, matches,
+						   matched_ind1, matched_ind2, unmatched_ind1, unmatched_ind2))
+		{
+			return false;
+		}
+		//drawing
+		Mat imMatches;
+		drawMatches(img1, prev_features_tmp, img2, curr_features_tmp, matches, imMatches);
+		imshow("matches2", imMatches);
+		waitKey();
+
+		prev_features.resize(matches.size());
+		curr_features.resize(matches.size());
+
+		std::vector<Point2f> points1(matches.size());
+		std::vector<Point2f> points2(matches.size());
+
+		Mat descriptors2(Size(32, matches.size()), CV_8U);
+
+		for (size_t m = 0; m < matches.size(); m++)
+		{
+			int i1 = matches[m].queryIdx;
+			int i2 = matches[m].trainIdx;
+
+			prev_features[m] = prev_features_tmp[i1];
+			curr_features[m] = curr_features_tmp[i1];
+
+			curr_descriptors_tmp.row(i2).copyTo(descriptors2.row(m));
+
+			points1[m] = prev_features[m].pt;
+			points2[m] = curr_features[m].pt;
+		}
+		curr_descriptors = descriptors2.clone();
+
+		Mat rvec(Mat::zeros(3, 1, CV_64F));   
+		Mat tvec(Mat::zeros(3, 1, CV_64F));    
+
+		solvePnPRansac(point_3f_frame[frames_number - 1], points2, camera_matrix, dist_coeffs, rvec, tvec);
+
+		Mat rmatrix;
+		Rodrigues(rvec, rmatrix);
+		R.push_back(rmatrix);
+		t.push_back(tvec);
+
+		frames_number++;
+		
+		return true;
+	}
+	else if (!is_map_init)
+	{
+		std::cout << "Map is not init" << std::endl;
+		return false;
+	}
+	else if (img2.empty())
+	{
+		std::cout << "Empty input image" << std::endl;
+		return false;
+	}
+}
+
+void Vslam::localMapping()
+{
+
+}
+
+void Vslam::detectFeatures(const Mat& img, std::vector<KeyPoint>& features, Mat& descriptors)
+{
+	Mat img_gray;
+	cvtColor(img, img_gray, cv::COLOR_BGR2GRAY);
+
+	int max_features = 2000;
+	Ptr<Feature2D> orb = ORB::create(max_features);
+	orb->detectAndCompute(img_gray, Mat(), features, descriptors);
```

Comment:
Features detection algorithm and parameters should be configurable.

Similar note is for matches.